working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+   type,
+
+4. AND the corresponding parameter in _A_ is implicitly typed,
+
+5. AND EITHER:
+
+   * The parameter in the invocation target corresponding to _B_ is function
+     typed, and _T_ is a free variable in its return type
+
+   * OR the parameter in the invocation target corresponding to _B_ is _not_
+     function typed, and _T_ is a free variable in its type.
+
+(The idea here is that we're trying to draw a dependency edge from _A_ to _B_ in
+precisely those cirumstances in which there is likely to be a benefit to
+performing a round of horizontal inference between type inferring _A_ and type
+inferring _B_).
+
+If the type of the invocation target is `dynamic` or the `Function` class, or
+some non-generic function type, then the resulting graph has no edges.
+
+So, for example, if the invocation in question is this:
+
+```dart
+f((t, u) { ... } /* A */, () { ... } /* B */, (v) { ... } /* C */, (u) { ... } /* D */)
+```
+
+And the target of the invocation is declared like this:
+
+```dart
+void f<T, U, V>(
+    void Function(T, U) a,
+    T b,
+    U Function(V) c,
+    V Function(U) d) => ...;
+```
+
+then the resulting dependency graph looks like this:
+
+&emsp;B &lArr; A &rArr; C &hArr; D
+
+(That is, there are four edges, one from A to B, one from A to C, one from C to D, and one from D to C).
+
+#### Stage selection
+
+After building the dependency graph, we condense it into its strongly connected
+components (a.k.a. "dependency cycles").  The resulting condensation is
+[guaranteed to be acyclic] (that is, the nodes in each strongly connected
+component depend transitively on each other and possibly on the nodes in other
+strongly connected components, but no cycle exists among multiple strongly
+connected components).  For the example above, C and D are condensed into a
+single strongly connected component, so the graph now looks like this:
+
+[guaranteed to be acyclic]: https://en.wikipedia.org/wiki/Strongly_connected_component#Definitions
+
+&emsp;(B) &lArr; (A) &rArr; (C &hArr; D)
+
+Then, for each stage of inference, we find all the strongly connected components
+with no graph edges pointing to them, and remove them from the graph.  We repeat
+the process for successive stages, until no strongly connected components remain
+in the graph.  The set of arguments we will recursively infer during each stage
+is the union of the arguments in all the strongly connected components removed
+from the graph during that stage.
+
+In this example, that means that there will be two stages.  The first stage will
+consist of argument A, and the second stage will consist of arguments B, C, and
+D.
+
+(Note that the invariant mentioned earlier, that non-function literals are
+always placed in the first stage, is guaranteed by the fact that dependency
+analysis only draws an edge from _A_ to _B_ if _A_ is a function literal.)
+
+### What constitutes a function literal?
```

Comment:
Done.  I decided I wasn't comfortable with the phrase "function literal argument" because the recursive nature of the definition means that we would be calling some things "function literal arguments" that are not, in fact, arguments at all, so I instead chose the phrase "function literal expression".

Reply:
Works!